Instance Metadata Service: Initial Commit

rifrankl opened the pull request:
Initial commit of Instance Metadata Service

The Azure Instance Metadata Service provides information about running virtual machine instances that can be used to manage and configure them.

-Added Swagger for 2018-10-01
-Added examples for 2018-10-01

### Latest improvements:
<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow. 
</i><br>
### Contribution checklist:
- [x] I have reviewed the [documentation](https://github.com/Azure/azure-rest-api-specs#basics) for the workflow.
- [x] [Validation tools](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist) were run on swagger spec(s) and have all been fixed in this PR.
- [x] The [OpenAPI Hub](https://aka.ms/openapiportal) was used for checking validation status and next steps.

rifrankl committed:
Instance Metadata Service: Initial Commit

Initial commit of Instance Metadata Service

-Added Swagger for 2018-10-01
-Added examples for 2018-10-01

AutorestCI commented:
# Automation for azure-sdk-for-go

AutorestCI commented:
# Automation for azure-sdk-for-js

AutorestCI commented:
# Automation for azure-sdk-for-node

AutorestCI commented:
# Automation for azure-sdk-for-ruby

AutorestCI commented:
# Automation for azure-sdk-for-python

AutorestCI commented:
# Automation for azure-sdk-for-java

azuresdkci commented:
Can one of the admins verify this patch?

salameer commented:
reassinging since @sarangan12  is oof

@rifrankl please fix the validation errors. you can check them through the [OpenAPI Hub here](https://portal.azure-devex-tools.com/app/branch/rifrankl/azure-rest-api-specs/rifrankl-imds-Microsoft.InstanceMetadataService-2018-10-01?source=github)

amarzavery commented on specification/imds/resource-manager/Microsoft.InstanceMetadataService/stable/2018-10-01/imds.json:
```diff
@@ -0,0 +1,725 @@
+                "description": "Get Instance Metadata for the Virtual Machine.",
+                "parameters": [
+                    {
+                        "name": "api-version",
```
api-version should be a global parameter definition

amarzavery commented on specification/imds/resource-manager/Microsoft.InstanceMetadataService/stable/2018-10-01/imds.json:
```diff
@@ -0,0 +1,725 @@
+        "ErrorResponse": {
+            "type": "object",
+            "properties": {
+                "error": {
```
The Error response does not follow the ARM guidelines for Error definition.
It must follow this shape as described [here](https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/common-api-details.md#error-response-content).

amarzavery replied:
@ravbhatnagar - Can you comment on this?

KrisBash replied:
This is not an ARM API. It should be moved to a /data-plane subdirectory in the repo. Once that is done, we can remove ARM sign-off requirements

rifrankl replied:
Do you mean simply moving the files to imds/data-plane from imds/resource-manager?

rifrankl replied:
Done.

amarzavery commented:
Please add language specific reademe.<language>.md files for go, nodejs, ruby, typescript, python. 

Take a look at [this](https://github.com/Azure/azure-rest-api-specs/tree/master/specification/storage/resource-manager) example.

amarzavery commented:
@rifrankl - Was this approved by ARM team? If not then you need to get this approved from ARM first and make sure you fix all the ARM linter errors.

rifrankl committed:
Instance Metadata Service: Updates

-Removed erroneous format tags
-Added security section

rifrankl committed:
Instance Metadata Service: Updates

-Moved to imds/data-plane from imds/resource-provider

rifrankl committed:
Instance Metadata Service: Added language .md files

amarzavery commented on specification/imds/data-plane/Microsoft.InstanceMetadataService/stable/2018-10-01/imds.json:
```diff
@@ -0,0 +1,731 @@
+                            "$ref": "#/definitions/attested"
+                        }
+                    },
+                    "400": {
```
Please replace all the negative response codes with only one response code. `"default"`.

```json5
"default": {
  "description": " Error response",
  "schema": {
     "$ref": "#/definitions/ErrorResponse"
   }
}
```
This will ensure that the sdk throws an exception for all the negative responses where the deserialized Error is based on the provided ErrorResponse schema.

amarzavery replied:
Please do this for all the operations.

rifrankl replied:
Done.

amarzavery commented on specification/imds/data-plane/Microsoft.InstanceMetadataService/stable/2018-10-01/imds.json:
```diff
@@ -0,0 +1,731 @@
+    "paths": {
+        "/instance": {
+            "get": {
+                "operationId": "getInstance",
```
We recommend operationId to be in the "Noun_Verb" format. This makes it easier for developers using the sdk to find logical operation groups on the client and then methods/actions that can be performed on those logical entities. Currently you would see all the methods on the client. This provides a subpar developer experience.

amarzavery replied:
operationId should be "Instances_Get". The reason to make the Noun plural is that "there is already a model definition named "Instance". Pluralizing operation groups resolves the naming conflict.

rifrankl replied:
Operations have been renamed to Instance_GetMetadata, Attested_GetDocument, Identity_GetToken, and Identity_GetInfo.

amarzavery commented on specification/imds/data-plane/Microsoft.InstanceMetadataService/stable/2018-10-01/imds.json:
```diff
@@ -0,0 +1,731 @@
+                        "type": "string",
+                        "description": "This is the version of API to invoke."
+                    },
+                    {
```
names of the parameter should be camelCase

rifrankl replied:
Move api-version to "parameters" section, under "ApiVersionParameter"

amarzavery commented on specification/imds/data-plane/Microsoft.InstanceMetadataService/stable/2018-10-01/imds.json:
```diff
@@ -0,0 +1,731 @@
+                }
+            }
+        },
+        "instance": {
```
Names of the model definition must be in "PascalCase". -> "Instance"

amarzavery replied:
Please add description for every model definition and for every property on the model definition. The description translates to REST API docs and SDK API docs on docs.microsoft.com website.

rifrankl replied:
Changed to PascalCase and added descriptions.

amarzavery commented on specification/imds/data-plane/Microsoft.InstanceMetadataService/stable/2018-10-01/imds.json:
```diff
@@ -0,0 +1,731 @@
+                }
+            }
+        },
+        "attested": {
```
PascalCase name and provide meaningful description

rifrankl replied:
Changed to Attested and added description

amarzavery commented on specification/imds/data-plane/Microsoft.InstanceMetadataService/stable/2018-10-01/imds.json:
```diff
@@ -0,0 +1,731 @@
+                        "not_found",
+                        "method_not_allowed",
+                        "too_many_requests"
+                    ]
```
Please add the extension ["x-ms-enum"](https://github.com/Azure/autorest/tree/master/docs/extensions#x-ms-enum). I think the extension will have `modelAsString: true`. This will ensure that an expandable enum is created in strongly typed languages, thereby avoiding breaking changes when a new value is added/removed to/from the expandable enum.

amarzavery commented on specification/imds/data-plane/Microsoft.InstanceMetadataService/stable/2018-10-01/examples/getattesteddata.json:
```diff
@@ -0,0 +1,57 @@
+{
+    "parameters": {
+        "api-version": "2018-10-01",
+        "nonce": "1111111111",
```
nonce is supposed to be a number and not string as per the schema definition.

rifrankl replied:
Corrected this. Nonce can be up to 32 alphanumeric characters.

amarzavery commented on specification/imds/data-plane/Microsoft.InstanceMetadataService/stable/2018-10-01/imds.json:
```diff
@@ -0,0 +1,731 @@
+        "description": "The Azure Instance Metadata Client",
+        "version": "2018-10-01"
+    },
+    "host": "169.254.169.254",
```
the host is usually a qualified name. I don't expect stable api to be called via ip address.

rifrankl replied:
This is how our service is used; that is, the user will make a request to http://169.254.169.254. For example, cloud init queries our service as follows:

open 'http://169.254.169.254/metadata/reprovisiondata?api-version=2017-04-02' with {'method': 'GET', 'url': 'http://169.254.169.254/metadata/reprovisiondata?api-version=2017-04-02', 'timeout': 1.0, 'allow_redirects': True, 'headers': {'Metadata': 'true', 'User-Agent': 'Cloud-Init/18.4-0ubuntu1~16.04.2'}} configuration

amarzavery commented on specification/imds/data-plane/Microsoft.InstanceMetadataService/stable/2018-10-01/imds.json:
```diff
@@ -0,0 +1,731 @@
+        },
+        "/attested/document": {
+            "get": {
+                "operationId": "getAttested",
```
AttestedData_Get

rifrankl committed:
Instance Metadata Service: Updated ErrorResponse

-Removed individual error codes from operations

amarzavery commented on specification/imds/data-plane/Microsoft.InstanceMetadataService/stable/2018-10-01/imds.json:
```diff
@@ -0,0 +1,731 @@
+                "description": "Get Attested Data for the Virtual Machine.",
+                "parameters": [
+                    {
+                        "name": "api-version",
```
Usually the api-version is defined as a global parameter (top level parameter property in the swagger spec). It has a special meaning to it. All the parameters specified in the the global parameters section will end up as properties on the generated client, unless specified to not do so via ["x-ms-parameter-location"](https://github.com/Azure/autorest/tree/master/docs/extensions#x-ms-parameter-location) extension.

rifrankl replied:
Moved api-version to "parameters" section, under "ApiVersionParameter"

amarzavery commented on specification/imds/data-plane/Microsoft.InstanceMetadataService/stable/2018-10-01/imds.json:
```diff
@@ -0,0 +1,731 @@
+                    "Get token"
+                ],
+                "summary": "Get a Token from Azure AD",
+                "description": "",
```
Empty description not allowed.

rifrankl replied:
Updated descriptions for every property.

amarzavery commented on specification/imds/data-plane/Microsoft.InstanceMetadataService/stable/2018-10-01/imds.json:
```diff
@@ -0,0 +1,731 @@
+                ],
+                "summary": "Get a Token from Azure AD",
+                "description": "",
+                "operationId": "getToken",
```
Identity_GetToken

rifrankl replied:
Done.

amarzavery commented on specification/imds/data-plane/Microsoft.InstanceMetadataService/stable/2018-10-01/imds.json:
```diff
@@ -0,0 +1,731 @@
+                        "type": "string",
+                        "enum": [
+                            "true"
+                        ],
```
Please use the x-ms-enum extension here. Using the x-ms-enum extension when the enum only has one value and the parameter/property is required has a different meaning to it. More info over [here](https://github.com/Azure/autorest/tree/master/docs/extensions#single-value-enum-as-a-constant).

amarzavery commented on specification/imds/data-plane/Microsoft.InstanceMetadataService/stable/2018-10-01/imds.json:
```diff
@@ -0,0 +1,731 @@
+                    "Get metadata information"
+                ],
+                "summary": "Get information about AAD Metadata",
+                "description": "",
```
no empty description.

amarzavery commented on specification/imds/data-plane/Microsoft.InstanceMetadataService/stable/2018-10-01/imds.json:
```diff
@@ -0,0 +1,731 @@
+                ],
+                "summary": "Get information about AAD Metadata",
+                "description": "",
+                "operationId": "getInfo",
```
Identity_GetInfo

amarzavery commented on specification/imds/data-plane/Microsoft.InstanceMetadataService/stable/2018-10-01/imds.json:
```diff
@@ -0,0 +1,731 @@
+                        "enum": [
+                            "2018-02-01",
+                            "2018-04-02",
+                            "2018-10-01"
```
Usually the api-version for all the apis in a given document remains constant. Why does this api-version parameter have an enum whereas others do not?

rifrankl replied:
Added enum to other instances of the parameter.

amarzavery commented on specification/imds/data-plane/Microsoft.InstanceMetadataService/stable/2018-10-01/imds.json:
```diff
@@ -647,29 +527,16 @@
             "properties": {
-                "error": {
-                    "type": "string",
+                "code": {
```
Are you sure this is how it looks on the wire? Swagger spec needs to correctly and completely describe the API as it is on the wire. Just want to make sure that this is the right thing. Had we merged the PR with the previous change and if you would have sent another PR with this change then that would result in a breaking change in the SDK.

Hence please be very careful about how you describe the APIs in the swagger spec.

rifrankl committed:
Instance Metadata Service: Updates to Swagger

-Created separate Identity error property
-Updated descriptions and added them where necessary
-Created parameters section and moved api-version and Metadata to it
-Renamed operationIds

rifrankl committed:
Added x-ms-parameter-location to Metadata parameter

amarzavery commented on specification/imds/data-plane/Microsoft.InstanceMetadataService/stable/2018-10-01/imds.json:
```diff
@@ -0,0 +1,760 @@
+    "paths": {
+        "/instance": {
+            "get": {
+                "operationId": "Instance_GetMetadata",
```
This should be Instances_GetMetadata

> OperationId has a noun that conflicts with one of the model names in definitions section. The model name will be disambiguated to 'InstanceModel'. Consider using the plural form of 'Instance' to avoid this. Note: If you have already shipped an SDK on top of this spec, fixing this warning may introduce a breaking change.

amarzavery commented on specification/imds/data-plane/Microsoft.InstanceMetadataService/stable/2018-10-01/imds.json:
```diff
@@ -0,0 +1,760 @@
+                    "200": {
+                        "description": "OK",
+                        "schema": {
+                            "$ref": "#/definitions/Attested"
```
Would recommend changing the name of the model to AttestedData to avoid naming conflict in model and operation group in the generated sdk

amarzavery commented on specification/imds/data-plane/Microsoft.InstanceMetadataService/stable/2018-10-01/imds.json:
```diff
@@ -0,0 +1,760 @@
+            "description": "Network Metadata"
+        },
+        "NetworkInterface": {
+            "type": "object",
```
missing description. Please make sure every model, property, operation, parameter has correct and complete description. Spec description translates to api doc on docs.microsoft.com and sdk api docs as well. Missing descriptions end up providing a sub par developer experience.

amarzavery commented on specification/imds/data-plane/Microsoft.InstanceMetadataService/stable/2018-10-01/imds.json:
```diff
@@ -0,0 +1,760 @@
+            ],
+            "description": "This must be set to 'true'.",
+            "required": true,
+            "x-ms-enum": {
```
x-ms-enum must not be used for constants. Please remove it.

amarzavery commented on specification/imds/data-plane/Microsoft.InstanceMetadataService/stable/2018-10-01/imds.json:
```diff
@@ -0,0 +1,760 @@
+            "enum": [
+                "2018-02-01",
+                "2018-04-02",
+                "2018-10-01"
```
We have one swagger per api version. What confuses me is that, you have the same apis for the three api versions.

If nothing changed between those api versions then why was the api version bumped?


FYI,
- The api version in the folder path and that in the top level info.version should be same.
- The api version in the top level info.version is used in sdk generation (client.apiVersion). It is a constant property and cannot be changed.

rifrankl committed:
-Minor property name changes
-Removed previous API versions from enum

nschonni commented on specification/imds/data-plane/Microsoft.InstanceMetadataService/stable/2018-10-01/imds.json:
```diff
@@ -0,0 +1,755 @@
+                    "description": "Error code",
+                    "enum": [
+                        "invalid_request",
+                        "unauthorize_client",
```
unauthorize_client -> unauthorized_client

rifrankl committed:
-Fixed a typo

rifrankl committed:
-Added missing descriptions

rifrankl committed:
-Added more missing descriptions

rifrankl commented:
The remaining errors are errors for which we would like an exception. See email thread from 11/20

rifrankl committed:
-Changed openapi type to data-plane

rifrankl committed:
-Added DefinitionsPropertiesNamesCamelCase suppression to readme

rifrankl committed:
-Forgot one camelcase suppression

amarzavery commented on specification/imds/data-plane/readme.md:
```diff
@@ -0,0 +1,74 @@
+       - $.definitions.IdentityTokenResponse.properties.object_id
+       - $.definitions.IdentityTokenResponse.properties.msi_res_id
+       - $.definitions.IdentityErrorResponse.properties.error_description
+	   
```
Can you remove the empty line here before the  \`\`\`?

I guess the linter error is failing for one property because of that, even though it has been added to the suppression list
-- | -- | --
R3016 - DefinitionsPropertiesNamesCamelCase | 2018-10-01/​imds.json:675 | Property named: 'error_description', for definition: 'IdentityErrorResponse' must follow camelCase style. Example: 'error_description'.
-- | -- | --

rifrankl committed:
-Removed a newline
